src/dex/synthetix/synthetix-state.ts

```diff
@@ -0,0 +1,933 @@
+        this.fullState.blockNumber === blockNumber) ||
+      // timeStamp condition
+      (timeStampInS &&
+        this.fullState &&
+        this.fullState.updatedAtInS < timeStampInS)
+    ) {
+      return this.fullState?.values;
+    }
+    return undefined;
+  }
+
+  async getOnchainState(blockNumber?: number): Promise<PoolState> {
+    if (this.onchainConfigValues === undefined)
+      throw new Error(
+        `${this.dexKey} is not initialized, but received pricing request`,
+      );
+
+    const addressesFromPK = this.onchainConfigValues.poolKeys.map(pk =>
+      dexPriceAggregatorUniswapV3.getPoolForRoute(
+        this.onchainConfigValues!.dexPriceAggregator.uniswapV3Factory,
+        this.onchainConfigValues!.dexPriceAggregator.overriddenPoolForRoute,
+        pk,
+      ),
+    );
+
+    const aggregatorAddressesWithoutZeros = Object.entries(
+      this.onchainConfigValues.aggregatorsAddresses,
+    ).reduce<Record<string, Address>>((acc, curr) => {
+      const [key, address] = curr;
+      if (address !== NULL_ADDRESS) {
+        acc[key] = address;
+      }
+      return acc;
+    }, {});
+
+    const [packCounter, slot0TickCumulativesSuspensionsLatestRoundCallData] =
+      this._buildObserveSlot0SuspensionsLatestRoundCallData(
+        addressesFromPK,
+        aggregatorAddressesWithoutZeros,
+      );
+
+    const slot0TickCumulativesSuspensionsLatestRound = (
+      await this.multiWrapper.tryAggregate<
+        Record<0 | 1, bigint> | Slot0 | boolean[] | boolean | LatestRoundData
+      >(true, slot0TickCumulativesSuspensionsLatestRoundCallData, blockNumber)
+    ).map(d => d.returnData) as (
+      | Record<0 | 1, bigint>
+      | boolean
+      | boolean[]
+      | LatestRoundData
+    )[];
+
+    let addressesFromPKBoundary = addressesFromPK.length * packCounter;
+    const slot0AndTickCumulatives =
+      slot0TickCumulativesSuspensionsLatestRound.slice(
+        0,
+        addressesFromPKBoundary,
+      );
+
+    const suspensions = slot0TickCumulativesSuspensionsLatestRound.slice(
+      addressesFromPKBoundary,
+      // Number of suspension requests
+      addressesFromPKBoundary + 3,
+    );
+
+    const latestRoundDatas = slot0TickCumulativesSuspensionsLatestRound.slice(
+      addressesFromPKBoundary + 3,
+    ) as LatestRoundData[];
+
+    const isSystemSuspended = suspensions[0] as boolean;
+    const synthSuspensions = suspensions[1] as boolean[];
+    const synthExchangeSuspensions = suspensions[2] as boolean[];
+
+    const areSynthsSuspended = Object.keys(
+      this.onchainConfigValues.addressToKey,
+    ).reduce<Record<string, boolean>>((acc, curr, i) => {
+      acc[curr] =
+        synthSuspensions[i] === true || synthExchangeSuspensions[i] === true;
+      return acc;
+    }, {});
+
+    const uniswapV3Slot0: Record<Address, Slot0> = {};
+    const tickCumulatives: Record<Address, Record<0 | 1, bigint>> = {};
+
+    _.chunk(slot0AndTickCumulatives, packCounter).map((result, i) => {
```

Comment:
Use `forEach` instead of `map` here?

Reply:
Your question lead me to google:
https://leanylabs.com/blog/js-forEach-map-reduce-vs-for-for_of/

And looks like `forEach` is almost two times faster than `map`. And `for of` is the fastest one. I didn't know about.
Fixed.

I always use `map` as rule of thumb, but shouldn't.

Reply:
`map` is for when you need an array of the output from running the function, if you don't need that, you can use `forEach`.  It's quite annoying that it's so much slower than `for ... of` and other plain loops though, maybe it's better to use those when you can...